zephyr/test/vmh.c

```diff
@@ -17,8 +18,258 @@
+	test_vmh_multiple_allocs(heap, 16, 1024, 4096);
+	test_vmh_multiple_allocs(heap, 16, 4096, 8192);
+
+	/* Clean up the heap after testing */
+	int ret = vmh_free_heap(heap);
+
+	zassert_equal(ret, 0, "Failed to free heap after multiple allocations");
+}
+
+/* Test case for vmh_alloc and vmh_free */
+static void test_vmh_alloc_free(bool allocating_continuously)
+{
+	struct vmh_heap *heap =
+	vmh_init_heap(NULL, MEM_REG_ATTR_CORE_HEAP, 0, allocating_continuously);
+
+	zassert_not_null(heap, "Heap initialization failed");
+
+	test_vmh_alloc_free_no_check(heap, 512, true);
+	test_vmh_alloc_free_no_check(heap, 2048, true);
+	test_vmh_alloc_free_no_check(heap, sizeof(int), false);
+	test_vmh_alloc_free_no_check(heap, 0, false);
+
+	test_vmh_alloc_free_check(heap, 512, true);
+	test_vmh_alloc_free_check(heap, 2048, true);
+	test_vmh_alloc_free_check(heap, sizeof(int), false);
+	test_vmh_alloc_free_check(heap, 0, false);
+
+	int ret = vmh_free_heap(heap);
+
+	zassert_equal(ret, 0, "Failed to free heap");
+
+	/* Could add tests with configs for heaps*/
+}
+
+/* Test case for vmh_alloc and vmh_free with and without config */
+static void test_heap_creation(void)
+{
+	test_vmh_init_and_free_heap(MEM_REG_ATTR_CORE_HEAP, NULL, 0, false, true);
+
+	/* Try to setup with pre defined heap config */
+	struct vmh_heap_config config;
+
+	config.block_bundles_table[0].block_size = 4;
+
+	config.block_bundles_table[0].number_of_blocks = 80;
+
+	config.block_bundles_table[1].block_size = 4;
+
+	config.block_bundles_table[1].number_of_blocks = 80;
+
+	test_vmh_init_and_free_heap(MEM_REG_ATTR_CORE_HEAP, &config, 0, false, true);
+}
+
+/* Test case for alloc/free on configured heap */
+static void test_alloc_on_configured_heap(bool allocating_continuously)
+{
+
+	/* Try to setup with pre defined heap config */
+	struct vmh_heap_config config;
+
+	config.block_bundles_table[0].block_size = 20;
+
+	config.block_bundles_table[0].number_of_blocks = 80;
+
+	/* Create continuous allocation heap for success test */
+	struct vmh_heap *heap =
+	vmh_init_heap(NULL, MEM_REG_ATTR_CORE_HEAP, 0, allocating_continuously);
+
+	/* Will succeed on continuous and fail with single block alloc */
+	test_vmh_alloc_free_check(heap, 512, allocating_continuously);
+
+	int ret = vmh_free_heap(heap);
+
+	zassert_equal(ret, 0, "Failed to free heap");
+}
+
+/* Test cases for initializing heaps on all available regions */
+static void test_vmh_init_all_heaps(void)
+{
+	int num_regions = CONFIG_MP_MAX_NUM_CPUS + VIRTUAL_REGION_COUNT;
+	int i;
+	const struct sys_mm_drv_region *virtual_memory_region =
+	sys_mm_drv_query_memory_regions();
+
+	/* Test initializing all types of heaps */
+	for (i = 0; i < num_regions; i++) {
+
+		/* Test without continuous allocation */
+		test_vmh_init_and_free_heap(virtual_memory_region[i].attr, NULL, 0, false,
+		true);
+
+		/* Test if it fails when heap already exists */
+		test_vmh_init_and_free_heap(virtual_memory_region[i].attr, NULL, 0, true,
+		false);
```

Comment:
I don't see a single contiguous heap test which is supposed to succeed. In fact why isn't this supposed to succeed? The comment says "already exists," but the test frees the heap internally?

Reply:
good catch need to rewrite this test